core/schema/module.go

```diff
@@ -316,176 +333,841 @@ func (s *moduleSchema) functionWithArg(ctx context.Context, fn *core.Function, a
+	ctx context.Context,
+	ref *core.GitModuleSource,
+	args struct{},
+) (string, error) {
+	return ref.HTMLURL(), nil
+}
+
+/*
+Convert the given module source into a normalized form that is consistent in ID and source dir/subpath, with
+the directory containing dagger.json always present at the root of the returned Directory.
+
+This is needed for a few different cases:
+ 1. If a git module source contains dagger.json at a subdirectory of the repo, we need to "re-root" it so
+ 2. Local modules sources can have many different IDs for equivalent sources due to the fact that the client
+    can load them from arbitrary parent dirs from a filesystem.
+*/
+func (s *moduleSchema) normalizeSourceForModule(
+	ctx context.Context,
+	src dagql.Instance[*core.ModuleSource],
+) (
+	newSrc dagql.Instance[*core.ModuleSource],
+	newRootDir dagql.Instance[*core.Directory],
+	newSourceSubpath string,
+	rerr error,
+) {
+	sourceSubpath, err := src.Self.Subpath()
 	if err != nil {
-		return nil, err
+		return newSrc, newRootDir, "", fmt.Errorf("failed to get module source subpath: %w", err)
+	}
+	sourceSubpath = filepath.Join("/", sourceSubpath)
+
+	rootDirPath, foundModsCfg, err := findUpConfigDir(ctx, src.Self.RootDirectory.Self, sourceSubpath)
+	if err != nil {
+		return newSrc, newRootDir, "", fmt.Errorf("error while finding config dir: %w", err)
+	}
+	if !foundModsCfg {
+		// No dagger.json found (okay if this is an uninitialized module source dir). Default to the root.
+		rootDirPath = "/"
+	}
+
+	// Reposition the root of the dir to the subdir containing the actual config
+	srcSubdirRelPath, err := filepath.Rel(rootDirPath, sourceSubpath)
+	if err != nil {
+		return newSrc, newRootDir, "", fmt.Errorf("failed to get module source path relative to config: %w", err)
+	}
+
+	switch src.Self.Kind {
+	case core.ModuleSourceKindGit:
+		if rootDirPath == "/" {
+			// No need to re-root
+			return src, src.Self.RootDirectory, sourceSubpath, nil
+		}
+
+		// Git sources are inherently normalized in terms of ID, but the root directory still needs to be
+		// re-rooted for the module
+		err = s.dag.Select(ctx, src.Self.RootDirectory, &newRootDir, dagql.Selector{
+			Field: "directory",
+			Args: []dagql.NamedInput{
+				{Name: "path", Value: dagql.String(rootDirPath)},
+			},
+		})
+		if err != nil {
+			return newSrc, newRootDir, "", fmt.Errorf("failed to reroot config directory: %w", err)
+		}
+		return src, newRootDir, filepath.Join("/", srcSubdirRelPath), nil
+
+	case core.ModuleSourceKindLocal:
+		/*
+			Local sources are the tricky one since the root dir is just some arbitrary directory (usually a local dir
+			loaded from a client's host filesystem).
+
+			The only way to consistently normalize the ID is to use a blob source, making the source id'd by its
+			content hash. In the case where this is an already correctly rooted source from a loaded local dir,
+			this should be mostly a cheap no-op since the blob will already be cached.
```

Comment:
Makes sense. What happens for a local dependency of a Git module? Does that also get normalized to a blob? Seems like that case at least could rely on the ID, but not sure if it's complicated to handle / worth handling

Reply:
The idea was that callers should call `ModuleSource.resolveDependency` and then provide *that* to `Module.withDependencies`; `resolveDependency` handled this case efficiently and turned a local dep source w/ git source parent into a git source, so you don't end up blob-ifying unnecessarily.

However I did realize that the requirement for clients to do that was kind of unnecessary; I updated the implementation of `Module.withDependencies` to just internally do this for clients, so they don't have to think about that anymore and we get all the nice behavior of keeping source `git` as much as possible.